### PR TITLE
Add support for template literals in enums

### LINKED ISSUE #30962: Backtick literals in enumerations
**TypeScript Version:**  3.4.3

**Search Terms:** backticks / literals / enums

**Code**

```ts
enum Foo {
  hello = `foo`,
};
```

**Expected behavior:**

It should work.

This is especially important because the `backticks` mode of the `quotes` rule from ESLint breaks programs (cf https://github.com/yarnpkg/berry/pull/70 for an example). While it could be seen as an ESLint bug, I think it *should* be possible for TS to use raw template string interchangeably with regular strings as long as they don't contain variables.

**Actual behavior:**

It throws an error (`type "foo" is not assignable to type 'Foo'`).

**Playground Link:**

https://www.typescriptlang.org/play/index.html#src=enum%20Foo%20%7B%0D%0A%20%20%20%20%20%20%20%20hello%20%3D%20%60foo%60%2C%0D%0A%7D%3B%0D%0A

**Related Issues:**

I think fixing this should have been part of #21476, but it focused on the case of `"a" + "b"` rather than the simpler case of a literal without variables.

### PR DESCRIPTION
**Pull request by IllusionMH**
Fixes #30962

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -6252,8 +6252,12 @@ namespace ts {
             return links.declaredType;
         }
 
+        function isStringOrNoSubstitutionTemplateLiteral(expr: Node): boolean {
```
Comment (ajafff) [inline src/compiler/checker.ts]: IIRC there's already a `isStringLikeLiteral` function in utilities.ts

Comment (IllusionMH) [reply src/compiler/checker.ts]: Thanks for catching this.

**Comment by IllusionMH** [comment]
One small question:
If this one is accepted do I need to update [Handbook - Enums](https://www.typescriptlang.org/docs/handbook/enums.html#computed-and-constant-members) section?
It has line "a literal enum expression (basically a string literal or a numeric literal)". Should it include somethins like "basically a string literal, a numeric literal, or a template literal without placeholders" or it is unnecessary detail/incorrect statement?

**Comment by RyanCavanaugh** [comment]
@ahejlsberg just wanted to get a :+1: from you on this one before proceeding

**Comment by ahejlsberg** [comment]
You made this comment in the issue:

> NoSubstitutionTemplateLiteral should be as good as StringLiteral anywhere here IMO

I agree, but this PR narrowly handles string literals in enum initializers. There are multiple other places in the checker where we look for `StringLiteral` but not `NoSubstitutionTemplateLiteral`. It's my opinion we should do all of them or none of them.

**Comment by weswigham** [comment]
> It's my opinion we should do all of them or none of them.

Except ES6 imports. Because per spec they aren't allowed there. :(

**Comment by IllusionMH** [comment]
Oh, I was working in scope of initial issue so this PR only supports enums.
If there are no objections - I will prepare new PR to cover other places and replace this PR.

Just to clarify scopeof expected PR:
1. Updates only to checker
1. Add support for `NoSubstitutionTemplateLiteral` where `StringLiteral` is supported
1. Do not add `TemplateExpression`s at this point
1. Exceptions:
    > ES6 imports. Because per spec they aren't allowed there.
    
    Which is already and error "String literal expected."

Is it correct?

**Comment by IllusionMH** [comment]
I'm finishing my investigation of places where support for `NoSubstitutionTemplateLiteral` can be added.
Should I create separate issue with list of changes for discussion or draft PR with detailed description would be enough?

**Comment by arcanis** [comment]
Is there a blocker on this issue I could help with?

**Comment by IllusionMH** [comment]
I've prepared list of places that have different behavior for template literals and regular strings.
Hope that scope described in #31548 will be approved and I'll send new PR (will include this changes too) this week.

**Comment by IllusionMH** [comment]
Closing this one in favor of https://github.com/microsoft/TypeScript/pull/32064

**Pull request closed without merging**